Disable Java serialization by default, #22333

patriknw opened the pull request:
Opening draft so you can see where this is going.

Refs #22333

* akka.actor.allow-java-serialization = off
* Moved primitive (Long, Int, String, ByteString) serializers
  from akka-remote to akka-actor since they had no dependency
  and are useful also in local systems, e.g. persistence.
  * e.g. needed for persistence-tck
  * less allow-java-serialization=on in tests
* CborSerializable in Jackson/test module for ease of use
* JavaSerializable for Java serialization in tests, already in akka-testkit,
  but misconfigured
* Made tests pass
  * allow-java-serialization=on in akka-persistence
  * allow-java-serialization=on in classic remoting tests
  * JavaSerializable and CborSerializable in other remoting tests
* Added serialization for
  * Boolean
  * java.util.concurrent.TimeoutException, AskTimeoutException

Remaining:
* [x] Fix tests in remaining modules
* [x] Solution for exceptions that are sent to parent when child actor fails
  * Any type of exception can't be serialized without loosing information
  * MiscMessageSerializer already has some support for exceptions
  * probably enough if we translate "unknown" exceptions to some general `RemoteException` class
  * good to be able to use this for Status.Failure (failing ask requests)
* [x] Docs and migration guide
* [ ] Decide what to do with `serialize-messages` https://github.com/akka/akka/issues/24273

patriknw committed:
Disable Java serialization by default, #22333

* akka.actor.allow-java-serialization = off
* Moved primitive (Long, Int, String, ByteString) serializers
  from akka-remote to akka-actor since they had no dependency
  and are useful also in local systems, e.g. persistence.
  * e.g. needed for persistence-tck
  * less allow-java-serialization=on in tests
* CborSerializable in Jackson/test module for ease of use
* JavaSerializable for Java serialization in tests, already in akka-testkit,
  but misconfigured
* Made tests pass
  * allow-java-serialization=on in akka-persistence
  * allow-java-serialization=on in classic remoting tests
  * JavaSerializable and CborSerializable in other remoting tests
* Added serialization for
  * Boolean
  * java.util.concurrent.TimeoutException, AskTimeoutException
* support for testing serialization with the inmem journal
* utility to verifySerialization, in SerializationTestKit
* remove AccountExampleWithCommandHandlersInState becuase not possible to serialize State when it's not static
  * Effect() is factory in EventSourcedBehavior  class
* test the account examples
* SharedLeveldbJournal.configToEnableJavaSerializationForTest
* support for exceptions from remote deployed child actors
  * fallback to akka.remote.serialization.ThrowableNotSerializableException
    if exception is not serializable when wrapped in system messages from
    remote deployed child actors and Status.Failure messages
  * it's implemented in `WrappedPayloadSupport.payloadBuilder`
* update reference documentation
* serialize-messages=off in most places, separate ticket for
  improving or removing that feature
* migration guide, including description of rolling update

patriknw commented:
I have added a solution for the exceptions in 56d6c77

I have started a Jenkins job to help me track down remaining test failures, but I think I have tried all modules locally.

Ready for reviewing. I will add/change documentation in additional commits here.

patriknw commented on akka-remote/src/main/scala/akka/remote/serialization/MiscMessageSerializer.scala:
```diff
@@ -122,10 +122,20 @@ class MiscMessageSerializer(val system: ExtendedActorSystem) extends SerializerW
     builder.setMessage(ex.getMessage).setCause(payloadSupport.payloadBuilder(ex.getCause)).build().toByteArray
   }
 
+  private def serializeThrowableNotSerializableException(ex: ThrowableNotSerializableException): Array[Byte] = {
```
the `ThrowableNotSerializableException` should be backported to 2.5.x to support rolling update

patriknw commented on akka-actor/src/main/scala/akka/serialization/PrimitiveSerializers.scala:
```diff
@@ -0,0 +1,198 @@
+/**
+ * INTERNAL API: only public by configuration
+ */
+@InternalApi private[akka] final class BooleanSerializer(val system: ExtendedActorSystem)
```
BooleanSerializer should be backported to 2.5.x to support rolling updates. Not important to add that serializer but we had some tests sending booleans.

johanandren commented on akka-actor/src/main/scala/akka/serialization/Serialization.scala:
```diff
@@ -550,4 +550,19 @@ class Serialization(val system: ExtendedActorSystem) extends Extension {
+    val serializer = findSerializerFor(obj)
+    val manifest = Serializers.manifestFor(serializer, obj)
+    deserialize(bytes, serializer.identifier, manifest).get
+  }
```
Since it is only for tests, why can't it live only in test kit?

patriknw replied:
needed it in InMemJournal too, and we probably want it from both typed and untyped testkits.

johanandren replied:
The typed testkit depends on the untyped so it can just delegate?

patriknw replied:
I see no reason for having that coupling, created issue https://github.com/akka/akka/issues/27303

johanandren replied:
Ok, In that case I'd still prefer duplicating the logic to having testkit stuff in the non-testkit codebase.

patriknw replied:
ok

johanandren commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -185,6 +186,24 @@ final class ActorTestKit private[akka] (val name: String, val config: Config, se
+   *                       and if not an `AssertionError` is thrown
+   * @return the deserialized object
+   */
+  def verifySerialization[M](obj: M, assertEquality: Boolean): M = {
```
I think a single param one with default `assertEquality` will be what (Scala) users need 99% of the time, so would be nice with that as a bonus.

helena commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/javadsl/TestKitJunitResource.scala:
```diff
@@ -143,4 +143,9 @@ final class TestKitJunitResource(_kit: ActorTestKit) extends ExternalResource {
+  /**
+   * See corresponding method on [[ActorTestKit]]
+   */
+  def verifySerialization[M](obj: M, assertEquality: Boolean): M = testKit.verifySerialization(obj, assertEquality)
```
Just a style thing, line 211 `verifySerialization` above is split into two lines but could be one, and this is one. I'd pick one for a commit.

patriknw replied:
this has been fixed, moved to separate SerializationTestKit

helena commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -18,9 +18,10 @@ import akka.actor.testkit.typed.internal.ActorTestKitGuardian
 import scala.concurrent.Await
 import scala.concurrent.duration._
+
+import akka.serialization.SerializationExtension
```
messy import grouping

johanandren commented on akka-cluster-metrics/src/multi-jvm/scala/akka/cluster/metrics/sample/StatsMessages.scala:
```diff
@@ -4,8 +4,10 @@
-final case class JobFailed(reason: String)
+final case class StatsJob(text: String) extends CborSerializable
+final case class StatsResult(meanWordLength: Double) extends CborSerializable
+final case class JobFailed(reason: String) extends CborSerializable
```
Quite noisy with the marker everywhere. I guess for typed you'd mark the top message type rather though so not as bad?

patriknw replied:
exactly

helena commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -185,6 +186,24 @@ final class ActorTestKit private[akka] (val name: String, val config: Config, se
+    val result =
+      SerializationExtension(internalSystem.toUntyped).verifySerialization(obj.asInstanceOf[AnyRef]).asInstanceOf[M]
+    if (assertEquality && result != obj)
+      throw new AssertionError(s"verifySerialization expected $obj, but was $result")
```
```suggestion
      throw new AssertionError(s"Serialization verification expected $obj but was $result.")
```

The function name would be seen anyway.

johanandren reviewed (approved):
LGTM, added some comments

helena commented on akka-actor-tests/src/test/scala/akka/serialization/PrimitivesSerializationSpec.scala:
```diff
@@ -0,0 +1,178 @@
+object PrimitivesSerializationSpec {
+  val serializationTestOverrides =
+    """
+  """
```
```suggestion
  val serializationTestOverrides = ""
```

helena commented on akka-actor/src/main/resources/reference.conf:
```diff
@@ -705,13 +705,29 @@ akka {
-    allow-java-serialization = on
+    allow-java-serialization = off
+
+    # Log warnings when the default Java serialization is used to serialize messages.
```
Do we even want to use the phrase `default Java serialization` vs drop the default?

patriknw replied:
yes, should be dropped.
fixed

patriknw commented:
updated docs in 5f726a9

patriknw commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:
```diff
@@ -185,6 +186,24 @@ final class ActorTestKit private[akka] (val name: String, val config: Config, se
+    if (assertEquality && result != obj)
+      throw new AssertionError(s"verifySerialization expected $obj, but was $result")
+    result
+  }
```
I mostly added it because it's very useful when testing EventSourcedBehaviors, e.g. serialization of state (snapshot) isn't easy to simulate. Ok, I can place it in a new `SerializationTestKit` but would still be nice to access it via `ActorTestKit` to not have several things to bring in.

patriknw replied:
moved in e513a71

johanandren replied:
👍

patriknw commented on akka-actor/src/main/resources/reference.conf:
```diff
@@ -705,13 +705,29 @@ akka {
+      primitive-long = "akka.serialization.LongSerializer"
+      primitive-int = "akka.serialization.IntSerializer"
+      primitive-string = "akka.serialization.StringSerializer"
+      primitive-bytestring = "akka.serialization.ByteStringSerializer"
```
all these were previously in akka-remote (aside from the new BooleanSerializer).
Yeah, not primitive like in Java primitive, but...

patriknw commented on akka-cluster-sharding-typed/src/test/java/jdocs/akka/cluster/sharding/typed/AccountExampleWithCommandHandlersInState.java:
```diff
@@ -1,297 +0,0 @@
- * delegate to methods in the state classes - replies of various types, using ExpectingReply and
- * EventSourcedBehaviorWithEnforcedReplies
- */
-public interface AccountExampleWithCommandHandlersInState {
```
had to remove AccountExampleWithCommandHandlersInState because not possible to serialize State when it's not static
  * Effect() is factory in EventSourcedBehavior  class

chbatey commented on akka-actor/src/main/resources/reference.conf:
```diff
@@ -705,13 +705,29 @@ akka {
+    # To be used with the above warn-about-java-serializer-usage
+    # When warn-about-java-serializer-usage = on, and this warn-on-no-serialization-verification = off,
+    # warnings are suppressed for classes extending NoSerializationVerificationNeeded
+    # to reduce noize.
```
this isn't just noise, it is noize!!!

chbatey commented on akka-actor/src/main/resources/reference.conf:
```diff
@@ -705,13 +705,29 @@ akka {
+
+    # Log warnings when the Java serialization is used to serialize messages.
+    # Java serialization is not very performant and should not be used in production
+    # environments unless you don't care about performance. In that case
```
and security?

chbatey commented on akka-cluster-tools/src/test/scala/akka/cluster/singleton/ClusterSingletonRestart2Spec.scala:
```diff
@@ -42,6 +42,9 @@ class ClusterSingletonRestart2Spec extends AkkaSpec("""
     }
   }
+  akka.actor.serialization-bindings {
+    "akka.cluster.UniqueAddress" = jackson-cbor
```
why override this? Comment?

patriknw replied:
there is no serializer for UniqueAddress, not intended to be sent as a standalone message

comment added

chbatey replied:
ah ofc

chbatey commented on akka-cluster/src/multi-jvm/scala/akka/cluster/StreamRefSpec.scala:
```diff
@@ -43,7 +44,7 @@ object StreamRefSpec extends MultiNodeConfig {
 
-  case class RequestLogs(streamId: Int)
+  case class RequestLogs(streamId: Int) extends CborSerializable
   case class LogsOffer(streamId: Int, sourceRef: SourceRef[String])
```
this one isn't serialised?

patriknw replied:
oversight, and actually turns out serialization of stream refs are not well supported at all: #27304

chbatey commented on akka-cluster/src/multi-jvm/scala/akka/cluster/routing/ClusterConsistentHashingGroupSpec.scala:
```diff
@@ -16,8 +16,8 @@ import scala.concurrent.Await
 
-  case object Get
-  final case class Collected(messages: Set[Any])
+  case object Get extends JavaSerializable
```
Any reason for the ones we use Cbor vs java in tests for?

patriknw replied:
added comment: using Java serialization because of `Any` in `Collected` (don't want to spend time on rewriting test)

`Get` would work with Cbor but thought it was best to use same

chbatey commented on akka-docs/src/main/paradox/persistence.md:
```diff
@@ -1091,7 +1091,7 @@ it must add
 @@snip [PersistenceSerializerDocSpec.scala](/akka-docs/src/test/scala/docs/persistence/PersistenceSerializerDocSpec.scala) { #custom-serializer-config }
 
-to the application configuration. If not specified, a default serializer is used.
+to the application configuration. If not specified, an exception will be throw when trying to persist events or snapshots.
```
:+1:

chbatey commented on akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md:
```diff
@@ -183,6 +183,11 @@ Classic remoting is deprecated but can be used in `2.6.` Explicitly disable Arte
 
+## Java Serialization
+
+Consistent hashing is serializing the hash key, unless it's bytes or String. If you use consistent hashing router
```
First sentence reads a bit odd, maybe start with the problem / user feature affected e.g.

`When using a consistent hashing router keys that were not bytes or a String were previously using the Java serializer. `

patriknw replied:
Well, it wasn't using Java serialization specifically. It used whatever serializer that was bound, and that could have been JavaSerializer. Improved wording.

chbatey commented on akka-persistence/src/main/scala/akka/persistence/journal/leveldb/LeveldbJournal.scala:
```diff
@@ -153,4 +158,25 @@ object SharedLeveldbJournal {
+  def configToEnableJavaSerializationForTest: Config = {
+    ConfigFactory.parseString(s"""
+    akka.actor.serialization-bindings {
+      "akka.persistence.journal.AsyncWriteTarget$$WriteMessages" = java-test
```
are these messages ever serialized? WOuldn't journals serialize the thing inside?

patriknw replied:
it's when using the journal proxy: `SharedLeveldbJournal`

chbatey commented on akka-remote-tests/src/test/resources/reference.conf:
```diff
@@ -1,6 +1,5 @@
   actor {
-    serialize-messages = on
-    warn-about-java-serializer-usage = off
+    serialize-messages = off # FIXME try to enable again?
```
which messages aren't serializable?

patriknw replied:
I have given up on this, or rather, I don't think there is any point in trying to support `serialize-messages`, but don't want to do that here. It's ticket #26802

chbatey commented on akka-remote/src/main/java/akka/remote/ArteryControlFormats.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2018-2019 Lightbend Inc. <https://www.lightbend.com>
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
```
intended?

patriknw replied:
it's protobuf generated

chbatey commented on akka-remote/src/main/java/akka/remote/SystemMessageFormats.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2018-2019 Lightbend Inc. <https://www.lightbend.com>
+ * Copyright (C) 2019 Lightbend Inc. <https://www.lightbend.com>
```
intended?

chbatey commented on build.sbt:
```diff
@@ -36,41 +36,41 @@ def isScala213: Boolean = System.getProperty("akka.build.scalaVersion", "").star
-    streamTestsTck,
-    streamTyped,
-    testkit)
+  actor,
```
intended reformat?

patriknw replied:
that's what scalafmt decides it to be (format on save)

chbatey commented on build.sbt:
```diff
@@ -303,15 +307,27 @@ lazy val protobuf = akkaModule("akka-protobuf")
+    testkit % "test->test",
+    streamTestkit % "test",
+    protobuf,
+    jackson % "test->test")
```
surprised these are all "test->test" and not just "test". We use the jackson prod files for the tests

patriknw replied:
I added `akka.serialization.jackson.CborSerializable` (marker interface) in src/test/scala, and as far as I know this is the way to include test sources from another project in sbt

chbatey commented on akka-testkit/src/main/scala/akka/testkit/TestException.scala:
```diff
@@ -0,0 +1,12 @@
+/**
+ * A predefined exception that can be used in tests. It doesn't include a stack trace.
+ */
+final case class TestException(message: String) extends RuntimeException(message) with NoStackTrace
```
:+1:

chbatey reviewed (commented):
Looks good. I've had a skim over all of this. let me know if there are parts I should pay more attention to

patriknw commented on akka-actor-tests/src/test/scala/akka/actor/FunctionRefSpec.scala:
```diff
@@ -36,7 +36,9 @@ object FunctionRefSpec {
 
-class FunctionRefSpec extends AkkaSpec with ImplicitSender {
+class FunctionRefSpec extends AkkaSpec("""
+  akka.actor.allow-java-serialization = on
```
ok, I'll do another scan of where I have used it, but in general it's where I didn't want to spend more time on rewriting the tests. I think I have used it rather sparingly overall.

johanandren replied:
Ok, I thought it was more complicated than that, then having that note here is good enough for me.

johanandren replied:
Aaaand, already updated with comments NVM. 👍

patriknw commented:
incorporated feedback in 191075c and added migration guide, including description of rolling update in  de7e581

patriknw commented:
PLS BUILD

patriknw commented:
Squashed and rebased. Should be ready for merge now. All tests have been running successfully in repeat job. Let's run a few more rounds.

I'll backport new new serializiers to 2.5 to support rolling updates.

Here is the full commit comment, which should give a good overview of what was done:

Disable Java serialization by default, #22333

* akka.actor.allow-java-serialization = off
* Moved primitive (Long, Int, String, ByteString) serializers
  from akka-remote to akka-actor since they had no dependency
  and are useful also in local systems, e.g. persistence.
  * e.g. needed for persistence-tck
  * less allow-java-serialization=on in tests
* CborSerializable in Jackson/test module for ease of use
* JavaSerializable for Java serialization in tests, already in akka-testkit,
  but misconfigured
* Made tests pass
  * allow-java-serialization=on in akka-persistence
  * allow-java-serialization=on in classic remoting tests
  * JavaSerializable and CborSerializable in other remoting tests
* Added serialization for
  * Boolean
  * java.util.concurrent.TimeoutException, AskTimeoutException
* support for testing serialization with the inmem journal
* utility to verifySerialization, in SerializationTestKit
* remove AccountExampleWithCommandHandlersInState becuase not possible to serialize State when it's not static
  * Effect() is factory in EventSourcedBehavior  class
* test the account examples
* SharedLeveldbJournal.configToEnableJavaSerializationForTest
* support for exceptions from remote deployed child actors
  * fallback to akka.remote.serialization.ThrowableNotSerializableException
    if exception is not serializable when wrapped in system messages from
    remote deployed child actors and Status.Failure messages
  * it's implemented in `WrappedPayloadSupport.payloadBuilder`
* update reference documentation
* serialize-messages=off in most places, separate ticket for
  improving or removing that feature
* migration guide, including description of rolling update

chbatey commented on akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md:
```diff
@@ -193,6 +193,64 @@ Classic remoting is deprecated but can be used in `2.6.` Explicitly disable Arte
+
+### Java serialization in consistent hashing
+
+When using a consistent hashing router keys that were not bytes or a String are serialized.
```
:+1:

chbatey commented on akka-remote/src/test/scala/akka/remote/serialization/AllowJavaSerializationOffSpec.scala:
```diff
@@ -108,15 +121,17 @@ class AllowJavaSerializationOffSpec
-      SerializationExtension(system).findSerializerFor(new ProgrammaticJavaDummy).toBinary(new ProgrammaticJavaDummy)
-      // should not receive this one, it would have been java serialization!
-      p.expectNoMessage(100.millis)
+      // should not allow deserializatio, it would have been java serialization!
```
typo deserializatio

chbatey reviewed (approved):
LGTM!

patriknw committed:
fix 2.13 compiler error

patriknw commented on akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md:
```diff
@@ -193,6 +193,64 @@ Classic remoting is deprecated but can be used in `2.6.` Explicitly disable Arte
+
+The procedure to changing from Java serialization to Jackson would look like:
+
+1. Rolling update from 2.5.x to 2.6.0
```
We will state somewhere that we only support rolling update from latest 2.5.x to 2.6.0, but I can change this to 2.5.24 (or later).

patriknw commented on akka-docs/src/main/paradox/project/migration-guide-2.5.x-2.6.x.md:
````diff
@@ -193,6 +193,64 @@ Classic remoting is deprecated but can be used in `2.6.` Explicitly disable Arte
+
+For compatibility with older systems that rely on Java serialization it can be enabled with the following configuration:
+
+```ruby
````
we use it in plenty of places for conf snippets, don't know if it's better than without

patriknw commented on akka-remote/src/test/scala/akka/remote/serialization/SystemMessageSerializationSpec.scala:
```diff
@@ -61,7 +62,8 @@ class SystemMessageSerializationSpec extends AkkaSpec(PrimitivesSerializationSpe
     // ActorInitializationException has no proper equality
-    "serialize and de-serialize Create(Some(ex))" in {
+    "serialize and de-serialize Create(Some(exc))" in {
+      // JavaSerializable is JavaSerializable
```
funny, TestException is JavaSerializable

patriknw committed:
minor review feedback

patriknw commented:
🎆 🍾
